### ISSUE TITLE
Compiler crashes when face with particular usage of the `some` keyword in a closure type

### ISSUE THREAD

**Issue by jeremyabannister** [apple/swift#64799]
**Description**

I just stumbled upon the very bad reaction that I received from the compiler/Xcode upon writing this function:

```swift
func foo () {
    var test: ()->some Equatable = { ()->Int in 0 }
}
```

If I paste this function into a fresh package I get only one error (not associated with any line of code): `Command CompileSwift failed with a nonzero exit code`, which shows me no additional detail underneath it when I click on it. Additionally, Xcode displays the message `An internal error occurred. Source editor functionality is limited. Attempting to restore...`.

**Steps to reproduce**

Attempt to compile the above function.

**Expected behavior**

"Build Succeeded" or a compilation error.

**Environment**
- Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)

**Comment by AnthonyLatsis** [comment]
Please provide the Xcode version in the original report

**Comment by LucianoPAlmeida** [comment]
I can see a crash in 5.6 and 5.7, but it looks fixed in 5.8 and main latest snapshot https://godbolt.org/z/b7dzP54oa
@jeremyabannister Can you please verify and close?

**Comment by AnthonyLatsis** [comment]
I am pretty sure this was fixed in the attached PR.
